jq255.md

```diff
@@ -0,0 +1,1425 @@
+
+def private_key_decode(bb):
+    sk = scalar_decode(bb)
+    if sk is False:
+        return False
+    if sk == int_to_scalar(0):
+        return False
+    return sk
+```
+
+To *generate* a private key, a non-zero scalar should be chosen from a
+cryptographically secure source. If the `secure_random(m)` function
+returns `m` random bytes with uniform probability, then the following
+function produces a new private key:
+
+```python
+def private_key_generate():
+    while True:
+        kb = secure_random(32)
+        sk = int_to_scalar(int.from_bytes(kb, byteorder='little'))
+        if sk != int_to_scalar(0):
+            return sk
+```
+
+**Note:** This function generates a random 256-bit integer, then reduces
+it modulo `r` (through the `int_to_scalar()` function). Since the value
+of `r`, in both jq255e and jq255s, is very close to a power of two, then
+this process induces only negligible bias; in the general case, one
+would need to produce a source integer substantially larger than `r`
+(e.g. 384 bits), but this is not required for jq255e and jq255s. Using a
+source 256-bit integer means that the reduction modulo `r` can be
+implemented in a simple way, without requiring a generic big integer
+division routine.
+
+The `private_key_generate()` function is organized as a loop so that the
+returned private key is guaranteed not to be zero. However, the
+probability that a zero private key value is obtained is negligible;
+hence, only one iteration of the loop will be used in practice.
+
+#### Public Keys
+
+A *public key* is a non-identity group element. The public key `Q`
+corresponds to the private key `sk` such that `Q = sk*G`, with `G` being
+the conventional generator. A public key is decoded and encoded as a
+group element, with the additional provision that, upon decoding,
+the identity element is rejected.
+
+```python
+def public_key_encode(Q):
+    return group_encode(Q)
+
+def public_key_decode(bb):
+    Q = group_decode(bb)
+    if Q is False:
+        return False
+    (E, Z, U, T) = Q
+    if U == GFq(0):
+        return False
+    return Q
+```
+
+The public key corresponding to the private key `sk` can be computed by
+invoking `point_mul(G, sk)`, with `G` being the conventional generator
+element.
+
+### Signatures
+
+A **signature** is computed over a given input message, using a private
+key; the signature value has size exactly 48 bytes. The correctness of
+the signature value over a given message can be verified against the
+public key corresponding to the private key used for signing.
+
+Internally, a Schnorr signature process is used; the challenge value is
+computed with the [BLAKE2][] hash function. Specifically, the BLAKE2s
+function is used, with a 32-byte output (as will be described below,
+that 32-byte output is then truncated to 16 bytes, but the hash output
+length 32 is still used in the BLAKE2 parameter block).
+
+In this section, we use the following notations:
+
+  - The signer's private key is `sk`.
```

Comment:
Signing key?

Reply:
There's always some ambiguousness about the "sign" term, because it designates both the overall signature system, and the action of producing a signature. That's how you end up with functions called something like `sign_verify()` and `sign_sign()`, and that is ugly. Here I went for the traditional characterization of keys with their private/public status. If we want to name them for their intended use, we end up with "signature generation key" and "signature verification key", but then there's the unfortunate collision with key generation, and we have "signature generation key generation", which, in terms of ugliness, is even worse.